doc/developer-notes.md

```diff
@@ -460,7 +460,14 @@ other input.
   safely continue even if the assumption is violated. In debug builds it
   behaves like `Assert`/`assert` to notify developers and testers about
   nonfatal errors. In production it doesn't warn or log anything, though the
-  expression is always evaluated.
+  expression is always evaluated. However, if the compiler can prove that a
+  statement inside `Assume` is side-effect-free, it may optimize the call away,
```

Comment:
Interesting.  If `Assume` can be optimized out during production possibly, is there any reason to prefer `Assert`?

There's an assert [here](https://github.com/bitcoin/bitcoin/blob/998386d4462f5e06412303ba559791da83b913fb/src/wallet/coinselection.cpp#L159) in a coin-selection algo that is meant to be performant.  Maybe this should be changed to `Asume`.

Reply:
I tested however the benchmarks don't seem to be sensitive enough for this to make a difference.